### ISSUE TITLE
imgui_impl_wgpu.cpp not working on latest Chrome Canary due to dependance on automatic Render Pipeline Layout Generation

### ISSUE THREAD

**Issue by tonygrue** [ocornut/imgui#6112]
**Version/Branch of Dear ImGui:**
```
Dear ImGui 1.89.3 WIP (18923)
--------------------------------
sizeof(size_t): 4, sizeof(ImDrawIdx): 2, sizeof(ImDrawVert): 20
define: __cplusplus=201703
define: IMGUI_DISABLE_OBSOLETE_KEYIO
define: __GNUC__=4
define: __clang_version__=16.0.0 (https://chromium.googlesource.com/external/github.com/llvm/llvm-project a1b09fc8de510f68b972567fac8eeeb8d8b2c461)
define: IMGUI_HAS_VIEWPORT
define: IMGUI_HAS_DOCK
--------------------------------
io.BackendPlatformName: imgui_impl_glfw
io.BackendRendererName: imgui_impl_webgpu
io.ConfigFlags: 0x00000000
io.ConfigViewportsNoDecoration
io.ConfigMacOSXBehaviors
io.ConfigInputTextCursorBlink
io.ConfigWindowsResizeFromEdges
io.ConfigMemoryCompactTimer = 60.0
io.BackendFlags: 0x0000040E
 HasMouseCursors
 HasSetMousePos
 PlatformHasViewports
 RendererHasVtxOffset
--------------------------------
io.Fonts: 1 fonts, Flags: 0x00000000, TexSize: 512,64
io.DisplaySize: 1500.00,800.00
io.DisplayFramebufferScale: 1.00,1.00
--------------------------------
style.WindowPadding: 8.00,8.00
style.WindowBorderSize: 1.00
style.FramePadding: 4.00,3.00
style.FrameRounding: 0.00
style.FrameBorderSize: 0.00
style.ItemSpacing: 8.00,4.00
style.ItemInnerSpacing: 4.00,4.00
```

**Back-end/Renderer/Compiler/OS**

Back-ends: imgui_impl_wgpu.cpp + imgui_impl_glfw.cpp
Operating System: MacOS, Chrome Canary Version 111.0.5554.0 (Official Build) canary (x86_64)

**My Issue/Question:**

imgui_impl_wgpu.cpp depends on "automatic layout generation, specifically:
"graphics_pipeline_desc.layout = nullptr; // Use automatic layout generation"

As of the latest version of Chrome Canary, this doesn't not appear to be supported anymore. IIRC, older versions were giving a warning, but it has now become a blocking error.

The current error is "Uncaught TypeError: Failed to execute 'createRenderPipeline' on 'GPUDevice': Failed to read the 'layout' property from 'GPUPipelineDescriptorBase': Required member is undefined."

**Standalone, minimal, complete and verifiable example:** _(see https://github.com/ocornut/imgui/issues/2261)_
```
Any Imgui use in webgpu/chrome will fail.
```

If I effectively revert https://github.com/ocornut/imgui/commit/83bdfef8e08286a8863242597e6c9f1e5b207a53, the problem appears to be fixed.

Specifically if I:
1) Remove
```
graphics_pipeline_desc.layout = nullptr; // Use automatic layout generation
```

2) Before wgpuDeviceCreateRenderPipeline I add:
```
WGPUBindGroupLayoutEntry common_bg_layout_entries[2] = {};
    common_bg_layout_entries[0].binding = 0;
    common_bg_layout_entries[0].visibility = WGPUShaderStage_Vertex;
    common_bg_layout_entries[0].buffer.type = WGPUBufferBindingType_Uniform;
    common_bg_layout_entries[1].binding = 1;
    common_bg_layout_entries[1].visibility = WGPUShaderStage_Fragment;
    common_bg_layout_entries[1].sampler.type = WGPUSamplerBindingType_Filtering;

    WGPUBindGroupLayoutEntry image_bg_layout_entries[1] = {};
    image_bg_layout_entries[0].binding = 0;
    image_bg_layout_entries[0].visibility = WGPUShaderStage_Fragment;
    image_bg_layout_entries[0].texture.sampleType = WGPUTextureSampleType_Float;
    image_bg_layout_entries[0].texture.viewDimension = WGPUTextureViewDimension_2D;

    WGPUBindGroupLayoutDescriptor common_bg_layout_desc = {};
    common_bg_layout_desc.entryCount = 2;
    common_bg_layout_desc.entries = common_bg_layout_entries;

    WGPUBindGroupLayoutDescriptor image_bg_layout_desc = {};
    image_bg_layout_desc.entryCount = 1;
    image_bg_layout_desc.entries = image_bg_layout_entries;

    WGPUBindGroupLayout bg_layouts[2];
    bg_layouts[0] = wgpuDeviceCreateBindGroupLayout(g_wgpuDevice, &common_bg_layout_desc);
    bg_layouts[1] = wgpuDeviceCreateBindGroupLayout(g_wgpuDevice, &image_bg_layout_desc);

    WGPUPipelineLayoutDescriptor layout_desc = {};
    layout_desc.bindGroupLayoutCount = 2;
    layout_desc.bindGroupLayouts = bg_layouts;
    graphics_pipeline_desc.layout = wgpuDeviceCreatePipelineLayout(g_wgpuDevice, &layout_desc);

    WGPUPipelineLayoutDescriptor pipeline_layout_desc = {};
    pipeline_layout_desc.label = "imgui_impl_wgpu_pipeline_layout_desc";
    pipeline_layout_desc.bindGroupLayoutCount = sizeof(bg_layouts) / sizeof(WGPUBindGroupLayout);
    pipeline_layout_desc.bindGroupLayouts = bg_layouts;
    graphics_pipeline_desc.layout = wgpuDeviceCreatePipelineLayout(g_wgpuDevice, &pipeline_layout_desc);
```

3) Remove the lower code that redefined bg_layouts
```
    // Create resource bind group
    WGPUBindGroupLayout bg_layouts[2];
    bg_layouts[0] = wgpuRenderPipelineGetBindGroupLayout(g_pipelineState, 0);
    bg_layouts[1] = wgpuRenderPipelineGetBindGroupLayout(g_pipelineState, 1);
```

I would offer a pull request, but I'm not sure if this is the solution the maintainer would prefer (and I have to run at the moment) but I wanted to at least document this.

**Comment by ocornut** [comment]
Hello,
Thanks for this!

We don't have a clear maintainer for this and I am not a user of WebGPU myself, but PR are generally welcome as it is a more optimal way to express and test a patch. Tagging @bfierz, @Kangz who may have interest in this.

**Comment by bfierz** [comment]
@ocornut,
Thanks for tagging me. I would be happy to take over maintance of the wgpu backend. I will try to look at the bug.

**Comment by bfierz** [comment]
@tonygrue, thanks for pointing out the issue.
I've opened a PR essentially applying your suggestion, which is the right thing to do.

**Comment by ocornut** [comment]
Thank you @tonygrue and @bfierz !
Merged this as d73e328

**Comment by Kangz** [comment]
> As of the latest version of Chrome Canary, this doesn't not appear to be supported anymore. IIRC, older versions were giving a warning, but it has now become a blocking error.

It is still supported but the way to express this in JS has been changed with a deprecation warning for many months. Previously `layout: undefined` worked whereas now you need an explicit `layout: "auto"`. I removed support for the deprecated path last week but it seems we forgot to update emscripten. I pinged on issue https://github.com/emscripten-core/emscripten/issues/18178 again to get that resolved soon.
